compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -257,22 +257,33 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         source_ty: Ty<'tcx>,
         cast_ty: Ty<'tcx>,
     ) -> InterpResult<'tcx> {
+        let (data, old_meta) = match *self.read_immediate(src)? {
+            Immediate::Scalar(data) => (data.check_init()?, ScalarMaybeUninit::Uninit),
```

Comment:
Using `ScalarMaybeUninit::Uninit` for "no input vtable" seems a bit suspicious, but I think it is fine. Still, a brief comment would be good.
```suggestion
            // If the input ptr is thin, use `Uninit` for the old metadata.
            // `unsize_just_metadata` knows how to handle that.
            Immediate::Scalar(data) => (data.check_init()?, ScalarMaybeUninit::Uninit),
```